awsiot/greengrass_discovery.py

```diff
@@ -0,0 +1,182 @@
+# permissions and limitations under the License.
+
+from awscrt.http import HttpClientConnection, HttpRequest
+from awscrt import io
+from awscrt.io import ClientBootstrap, ClientTlsContext, TlsConnectionOptions, SocketOptions
+import awsiot
+from concurrent.futures import Future
+import json
+
+class DiscoveryClient(object):
+    __slots__ = ['_bootstrap', '_tls_context', '_socket_options', '_region', '_tls_connection_options', '_gg_server_name', 'gg_url', 'port']
+
+    def __init__(self, bootstrap, socket_options, tls_context, region):
+        assert bootstrap is not None and isinstance(bootstrap, ClientBootstrap)
+        assert socket_options is not None and isinstance(socket_options, SocketOptions)
+        assert tls_context is not None and isinstance(tls_context, ClientTlsContext)
+        assert region is not None and isinstance(region, str)
+
+        self._bootstrap = bootstrap
+        self._socket_options = socket_options
+        self._region = region
+        self._gg_server_name = 'greengrass-ats.iot.{}.amazonaws.com'.format(region)
+        self._tls_connection_options = tls_context.new_connection_options()
+        self._tls_connection_options.set_server_name(self._gg_server_name)
+        self.port = 8443
+
+        if io.is_alpn_available():
+            self._tls_connection_options.set_alpn_list('x-amzn-http-ca')
+            self.port = 443
+
+    def discover(self, thing_name):
+        ret_future = Future()
+        response_body = bytearray()
+        request = None
+
+        def on_incoming_body(response_chunk):
+            response_body.extend(response_chunk)
+
+        def on_request_complete(completion_future):
+            global request  
+            try:
+                response_code = request.response_code
+                # marking request as global prevents the GC from reclaiming it,
+                # so force it to do it here.
+                request = None
+                if response_code == 200:
+                    payload_str = response_body.decode('utf-8')
+                    discover_res = DiscoverResponse.from_payload(json.loads(payload_str))
+                    ret_future.set_result(discover_res)                    
+                else:                    
+                    ret_future.set_exception(DiscoveryException('Error during discover call: response code ={}'.format(response_code), response_code))
+
+            except Exception as e:
+                ret_future.set_exception(e)
+
+        def on_connection_completed(conn_future):
+            global request
+            try:
+                connection = conn_future.result()                
+                request = connection.make_request(
+                    method='GET', 
+                    uri_str='/greengrass/discover/thing/{}'.format(thing_name), 
+                    outgoing_headers={'host':self._gg_server_name}, 
+                    on_outgoing_body=None, 
```

Comment:
This is redundant isn't it? Doesn't that callback default to None?

Reply:
Well, it's explicit and clear, but also reduntant... I don't have that strong of an opinion on it.